Bug fix: Removing items from linksToMany relationships

burieberry opened the pull request.

burieberry committed:
Add failing test for linksToMany item removal

burieberry committed:
linksToMany removal bug fix

unskip test

lukemelia commented on packages/runtime-common/realm.ts:
```diff
@@ -594,6 +594,13 @@ export class Realm {
+      // relationships should be overwritten instead of merged, otherwise we
+      // won't be able to remove items from linksToMany relationships
+      card.data.relationships = patch.data.relationships;
+    }
```
Does the JSON-API spec have an opinion about this? This seems a little contrary to the idea of patch. I'd expect a null-value-aware merge would be appropriate here?

habdelra replied:
i think burcu got this idea from how we handle the contains-many fields. but yeah i agree, this does seem antithetical to how patches should work for both contains many and linksto many

burieberry replied:
yes, I got the idea from how we're handling the containsMany arrays above this code. I'll see how else we can handle these but I'm not sure... If there are 2 items and the first one is removed, the second item needs to become the first and only item. We don't want first item to show `null` and second item to display a link. If all items are removed, then we need to change the link to `null`. 

linksToMany with items:
```
relationships: {
            'pets.0': {
              links: {
                self: `${testRealmURL}dir/mango`,
              },
            },
            'pets.1': {
              links: {
                self: `${testRealmURL}dir/van-gogh`,
              },
            },
          },
```

linksToMany with no items:

```
relationships: {
            'pets': {
              links: {
                self: null,
              },
            },
          },
```

habdelra replied:
thinking more about it, I think we want the ability to have sparse arrays, as well as to be able to remove items from arrays. with @lukemelia's suggestion i don't think that is possible, as `null` would be signals to splice the array. i think we consider field's value to be the entire array. so if you want to patch a linksTo many field such that you want to remove items in the array, since the value of the field is an array, you need to provide the whole array in the patch. The idea here is that we are patching fields in the card, we are not patching below the level of a field. 

The containsMany is probably the messier example of this since you could have arrays of cards that have fields that are cards that could have arrays of more cards. Maybe a better idea here is that you could patch a card within the containsMany using the index number? (for linksToMany i think this never comes up since the deeper item is in the linked card which is not part of the consuming card's data)

burieberry replied:
JSON-API specs allow for a full replacement when doing a PATCH on a to-many relationship (https://jsonapi.org/format/#crud-updating-resource-relationships). Since our linksToMany relationships are not serialized in an array format (this was a choice on our end for various reasons), in this PR I added a middle step of converting linksToMany relationships to array format to be able to perform the merge correctly. After the merge, I convert it back to the non-array format.

burieberry replied:
Of course the previous solution was simpler, but it required passing in the full relationships object, instead of just the changed part. The current version allows for just passing in the change in the request (see the two tests I added in `realm-test`).

burieberry committed:
Better merge of patched relationship data

- Make linksToMany test more complex
- Added linksTo field to pet-person card and updated the places the card was used in tests

burieberry committed:
Await value in async test

lukemelia reviewed (approved):
Definitely addressed my concern. 👍🏼

burieberry committed:
more tests and bug fix

burieberry committed:
Merge remote-tracking branch 'origin/main' into cs-5612-removing-linksToMany-item
